### ISSUE TITLE
projectcalico.org/operator-node-migration label stuck in pre-operator on upgrade. 

### ISSUE THREAD

**Issue by paulgmiller** [projectcalico/calico#4525]
Calico upgrade failed as label was never updated

## Expected Behavior
Calico-node ds should install on all nodes upgrading calico

## Current Behavior
we had a cluster go from calico 3.7.2 to 3.17 on aks. However while the operator was running  nothig under calico-system was able to start. In particular kube-controller was pending and calico-node daemonset didn't like any of the nodes. 

Seems 
## Possible Solution
We evntually realized the deamonset had a node selector on projectcalico.org/operator-node-migration=migrated
All our nodes,  **including fresh nodes brought up after the upgrade** had projectcalico.org/operator-node-migration=pre-operator.

Manually overwriting the label made the daemonset schedule and everything came up just fine. 

## Steps to Reproduce (for bugs)
We don't have a good reproduction. When we tried this again we did not get stuck in the same way.

## Your Environment

* Calico version tryign to goto 3.17. 
* Orchestrator version (e.g. kubernetes, mesos, rkt):  azure kubernetes 
* Operating System and version: ubuntu 18. 
* Link to your project (optional):

**Comment by paulgmiller** [comment]
Lookng through the tigera operator logs I see this repeated
{"level":"error","ts":1618258406.7740195,"logger":"controller_installation","msg":"error migrating resources to calico-system","Request.Namespace":"","Request.Name":"default","error":"the kube-system node DaemonSet is not ready with the updated nodeSelector: daemonsets.apps \"calico-node\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/tigera/operator/pkg/controller/installation.(*ReconcileInstallation).SetDegraded\n

**Comment by tmjd** [comment]
What was happening with the kube-system calico-node DaemonSet? Were calico-node pods (in kube-system) not becoming ready? The operator waits for the daemonset to become ready before continuing with the migration, what was blocking those pods from becoming healthy and ready?

**Comment by mad01** [comment]
i got the same issue. my problem was that i had removed the old calico before applying the new so the tigera operator could not complete it's migration. After understanding that i added back the old and restarted the update again and the tigera operator was able to complete the migration. the old daemonset in kube-system got removed as part of the migration the operator did.

**Comment by paulgmiller** [comment]
The calico node daemonset doesn't schedule on any node because it has a
node.selector doe operator-node-migration=migrated

**Comment by tmjd** [comment]
The kube-system calico-node daemonset should have been updated by the operator to have a selector `projectcalico.org/operator-node-migration=pre-operator`, and then the operator waits for all the pods to become ready before continuing with the migration. The message suggests that the kube-system daemonset was removed
> the kube-system node DaemonSet is not ready with the updated nodeSelector: daemonsets.apps "calico-node" not found

**Comment by tmjd** [comment]
@paulgmiller did you manually remove the calico-node daemonset from the kube-system namespace at some point? Since you did not mention it I'm assuming you did not but I was hoping you could confirm that assumption.

**Comment by paulgmiller** [comment]
It was removed but only after it was already "stuck" and not scheduling on
any nodes. Operator put it back but we remained stuck. May be able to get
it's node selector before we deleted it from api server logs

**Comment by tmjd** [comment]
Ah ok. It would have been interesting to know what the issue was before the daemonset was removed. FYI if you run into issues with the upgrade again, removing the daemonset will definitely cause an issue. Could you collect the tigera-operator logs and add them to this issue?
I'd like to be able to review them to see what the original issue might have been.

**Comment by paulgmiller** [comment]
[tiger.log](https://github.com/projectcalico/calico/files/6352395/tiger.log)
Attached log. I could easily see a situation where upgrading the original nodes failed because they were accidentally cordoned. I was more suprised that a newly minted node vm that never had the old pre-operator daemonset on it got stuck in the same way.

**Comment by tmjd** [comment]
I believe the concern with the cordoned nodes should not be an issue if the calico-node daemonset in the kube-system namespace is not removed. I think cordoned nodes are accounted for when the operator checks that the daemonsets are up to date.
The issue here seems to be that the calico-node daemonset in the kube-system namespace was removed. You Suggested that the operator put back the calico-node daemonset but the operator never creates a daemonset in the `kube-system` namespace.

**Comment by tmjd** [comment]
Looking at the attached logs seems to indicate that the the `calico-node` daemonset in the `kube-system` namespace was removed before the operator was started. I'm not sure if that is because the operator was restarted so the logs don't include when it first started migration or if the `calico-node` deamonset in the `kube-system` namespace was removed when the operator was installed.

 > I was more suprised that a newly minted node vm that never had the old pre-operator daemonset on it got stuck in the same way.

This would be expected since when migrating the operator adds the node selector to ensure that it can control which nodes have the old (`kube-system`) and new (`calico-system`) calico-node pod.

In general when upgrading to the operator from a previous manifest based install, the operator goes through this procedure:
1. Operator detects that the previous install was a manifest based one, all the components for the manifest based one so it looks for `calico-node` daemonset, `calico-kube-controllers`, and possibly `calico-typha`, if any of them are in the `kube-system` namespace then the operator knows it needs to go through a migration.
2. The operator labels all the existing nodes with `projectcalico.org/operator-node-migration=pre-operator`
3. Updates the 'old'/manifest based calico-node daemonset with the node selector for the above label. Then the operator waits for the `calico-node` daemonset in `kube-system` to be fully rolled out. <== This seems to be where the operator was stuck.
4. The operator creates the new `calico-node` daemonset in the `calico-system` namespace, this includes the nodeSelector `projectcalico.org/operator-node-migration=migrated`. This is to ensure the operator can in a controlled fashion switch nodes one at a time from the `calico-node` daemonset from the `kube-system` to `calico-system` namespace.
5. When all nodes are switched, then the operator will cleanup:
   * delete the `calico-node` daemonset in the `kube-system` namespace
   * remove the nodeSelector from the `calico-node` daemonset in the `calico-system` namespace

**Comment by paulgmiller** [comment]
Discussing with Erik we use [addon manager](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/addon-manager) and reconcile to apply calico so it's quite possible the kube-system ripped out the old daemonet while the opetor was watching. Considering if ensure exists will mitgate this.

**Comment by mattstam** [comment]
For anyone experiencing an upgrade failure due to this, this should solve the issue:

1. Check to see if `tigera-operator` is running but the node still has the label `projectcalico.org/operator-node-migration=pre-operator`
2. If so, overwrite the label: `kubectl label node <node> projectcalico.org/operator-node-migration=migrated --overwrite`

That should get all the new calico-system resources running. In some cases there may still be old kube-system calico resources there, so you may want to manually delete those.

**Comment by paulgmiller** [comment]
Got another instance of this today with the 1.17.1 operator. Attaching logs but they're not that helpful.
[tigeraoperator.log](https://github.com/projectcalico/calico/files/6657472/tigeraoperator.log)

 Basically spins at 
    https://github.com/tigera/operator/blob/c60c32b736d3258bd475e4c74112b225998a3182/pkg/controller/migration/namespace_migration.go#L446
 
When it can't find the kube-system daemonset deleted by addon manager.  There are 5 restarts so the actual decision to migrate is in a log that's lost now (previous is the same as this just starts at 16:19:53).

**Comment by KannappanSomu** [comment]
there is similar issues in calico with aks , is there any ETA on the fix

**Comment by tmjd** [comment]
@KannappanSomu do you have any details details/information from your cluster when this happened?
The original reason for this issue was that AKS does not allow for the normal upgrade process to the operator installation so it was hitting an un-expected cases because the AKS process was removing the previous calico resources (in `kube-system`), this has been addressed by having the operator ignore those resources if they are in the process of being deleted. This updated version has been rolled out in AKS.
[The update in this comment](https://github.com/projectcalico/calico/issues/4525#issuecomment-861683591) suggests the current problem is that an old calico-typha was possibly being re-installed by addonmanager, I do not know if a fix for this has been rolled out though. And there is not much the Calico team can do to address a problem like that.

It is not very helpful to say there is a "similar issue", because we do not know how your issue is different without more information. If this happens again please collect information that includes:
* Nodes and their labels
* Deployments and daemonsets in the 'kube-system' namespace
* Deployments and daemonsets in the 'calico-system' namespace
* Output of `kubectl get tigerastatus -o yaml`
* Logs from the tigera-operator in the `tigera-operator` namespace
With this information we should be able to tell if the problem you experienced is "old" typha being reinstalled or a new issue to review.

**Comment by KannappanSomu** [comment]
@tmjd  I was able to fix by removing the IP address annotation assigned wrongly and restart calico nodes . 

this is explained here in detail 

https://github.com/projectcalico/calico/issues/4611

**Comment by nununoisy** [comment]
Was able to reproduce on AKS while upgrading 2 K8s clusters from 1.19.7 to 1.20.7. Unfortunately I can't seem to find any relevant logs that haven't already been seen.

I was curious as to how the migration process worked, in particular a slight discrepancy between the conditions required to successfully migrate and the conditions that cause the check to pass.

[CoreNamespaceMigration.NeedsCoreNamespaceMigration](https://github.com/tigera/operator/blob/f25d83f31bd92fef29306f8c4d114a9f00adb192/pkg/controller/migration/namespace_migration.go#L87) checks if any of the old namespaced resources exist. However, [CoreNamespaceMigration.Run](https://github.com/tigera/operator/blob/f25d83f31bd92fef29306f8c4d114a9f00adb192/pkg/controller/migration/namespace_migration.go#L217) will only succeed if all of the old namespaced resources exist. The discrepancy is normally impossible if tigera is the only service managing calico resources but this is not always the case (e.g. AddonManager in Reconcile mode). 

The question then shifts to which of the two we change to remove the discrepancy. We have two options:
- Modify NeedsCoreNamespaceMigration such that it requires all of the old components exist. I don't think this is a good idea since it would ignore fragmented/partial old installations.
- Modify Run and [CoreNamespaceMigration.ensureKubeSysNodeDaemonSetHasNodeSelectorAndIsReady](https://github.com/tigera/operator/blob/f25d83f31bd92fef29306f8c4d114a9f00adb192/pkg/controller/migration/namespace_migration.go#L444) such that it ignores resources that do not exist. This means partial installations are still migrated but also don't hang while migrating.

A possible fix might be to ignore NotFound errors when trying to modify/delete the DaemonSet since it was likely already deleted by something else. If that resource comes back after a partial migration chances are it wouldn't be Calico's fault (though if there is a case I'm not considering here please do let me know).

**Comment by tmjd** [comment]
@nununoisy why do you think the migration was hanging? My understanding is that AKS should be removing calico-node and typha from the kube-system namespace and because of that the operator should not be attempting to perform a migration. Even if it did initially detect that a migration was needed (because typha or node still existed) when it got to `ensureKubeSysNodeDaemonSetHasNodeSelectorAndIsReady` it should return an error if that daemonset did not exist which should cause the whole operator Reconcile to restart. 

I think the only reason it would stay stuck would be if the old typha or node were not removed.

@paulgmiller said
>UPDATE: we found an old calico-typha deployment hanging out in kube-system with addonmanager ensure exists rather than reconcile. That seems to have triggered the migration which then couldn't find calico node. We're not sure if other clusters have the same leftover so checking....

@nununoisy Could that have been the case that typha was being added back by AddonManager instead of being removed like expected?

**Comment by nununoisy** [comment]
Looking through Log Insights again I can see that a `calico-typha` instance was being created inside of `kube-system` around the time we were doing the upgrade, which seems to corroborate the issue in @paulgmiller's comment.

Interestingly, I found a strange issue when combing through `KubeEvents`.

```
Failed to pull image "mcr.microsoft.comoss/calico/typha:v3.19.0": rpc error: code = Unknown desc = failed to pull and unpack image "mcr.microsoft.comoss/calico/typha:v3.19.0": failed to resolve reference "mcr.microsoft.comoss/calico/typha:v3.19.0": failed to do request: Head https://mcr.microsoft.comoss/v2/calico/typha/manifests/v3.19.0: dial tcp: lookup mcr.microsoft.comoss on <ip>: no such host
```
Maybe this lingering instance is here because the new one fails to pull an image (it should also be noted that there is a PDB for `calico-typha` with a `maxUnavailable` of 1 which may be the reason behind the old instance being restarted). I think the issue may be that there's a typo in the container URL (it should probably be "mcr.microsoft.com/oss/calico/typha:v3.19.0"). I noticed the same issue for `calico/pod2daemon-flexvol:v3.19.0`.

What I may try is manually overwriting the `calico-typha` spec to fix the container URL and see if the old instance goes away.

**Comment by mattstam** [comment]
@tmjd so context for the above is that we found an issue in AKS where we had some old calico typha's were left around in kube-system. This was due to us implementing a workaround for a for scheme change that happened around September 2020. This consisted of us renaming the `calico-typha` deployment to `calico-typha-deployment` deployment. We've since released a change to purge any calico-typha deploy for < v1.19 clusters.

And for @nununoisy, this `comoss` in the URL was a bug we encountered on a few different clusters. We released a fix for this by changing the `registry` value in the Installation from `mcr.microsoft.com` to `mcr.microsoft.com/`.

**Comment by nununoisy** [comment]
I did notice the `calico-typha-deployment` name discrepancy but it didn’t click until just now. That’s good to hear - I’ll relay this back to the team.

**Comment by nununoisy** [comment]
After manually removing the old typha deploys the cluster now properly scales horizontally and the `projectcalico.org/operator-node-migration` label is gone as was expected from the upgrade.

I believe the question remains though - is this considered a bug? I would imagine a partial upgrade or lingering old components might happen (which did occur in this case) and that causes the upgrade to hang. Should those edge cases be handled, or at the very least could they be documented in case this happens to someone else looking to upgrade Calico?

**Comment by tmjd** [comment]
I don't think there is a bug here anymore, since the change has been added to ignore the resources if they're terminating, and as long as the expected [upgrade process](https://docs.projectcalico.org/maintenance/operator-migration) is followed then there shouldn't be an issue.
Having an incomplete install should be understandable from the messages that the operator will report in `kubectl get tigerastatus calico -o yaml`. If we aren't clear in the error messages that we only found some of the components needed then we should correct that. If this isn't clear then please create a bug on the tigera/operator repo.
